File: app/models/webapp/renew.py
"""WebApp DTOs: manual renewal options and confirmation."""

from pydantic import BaseModel, Field


class RenewPlanItem(BaseModel):
    """Single plan for renewal."""

    id: int
    storage: float
    duration: int
    price: int
    plan_name: str
    ip_limit: int = 0


class WebAppRenewOptionsRequest(BaseModel):
    """Request for renewal options (durations and plans)."""

    code: int = Field(..., description="Service code")
    session_token: str | None = Field(None, description="Session token from login")
    init_data: str | None = Field(None, description="Telegram init data")


class WebAppRenewOptionsResponse(BaseModel):
    """Response with renewal options."""

    ok: bool
    service_code: str | None = None
    panel_name: str | None = None
    is_fair_usage: bool = False
    durations: list[int] | None = None
    duration_groups: dict[str, list[int]] | None = None
    plans: list[RenewPlanItem] | None = None
    error: str | None = None


class WebAppRenewConfirmRequest(BaseModel):
    """Request to confirm renewal (with optional discount code)."""

    code: int = Field(..., description="Service code")
    plan_id: int = Field(..., description="Selected plan id")
    discount_code: str | None = Field(None, description="Optional discount code")
    session_token: str | None = Field(None, description="Session token from login")
    init_data: str | None = Field(None, description="Telegram init data")


class WebAppRenewConfirmResponse(BaseModel):
    """Response after renewal confirm."""

    ok: bool
    message: str | None = None
    new_balance: int | None = None
    new_volume: str | None = None
    amount_paid: int | None = None
    config_name: str | None = None
    error: str | None = None

File: frontend/src/types/webapp/renew.ts
/** Mirrors app/models/webapp/renew.py */
import type { WebAppAuthRequest } from "./common";

export interface RenewPlanItem {
  id: number;
  storage: number;
  duration: number;
  price: number;
  plan_name: string;
  ip_limit: number;
}

export interface WebAppRenewOptionsRequest extends WebAppAuthRequest {
  code: number;
}

export interface WebAppRenewOptionsResponse {
  ok: boolean;
  service_code?: string | null;
  panel_name?: string | null;
  is_fair_usage: boolean;
  durations?: number[] | null;
  duration_groups?: Record<string, number[]> | null;
  plans?: RenewPlanItem[] | null;
  error?: string | null;
}

export interface WebAppRenewConfirmRequest extends WebAppAuthRequest {
  code: number;
  plan_id: number;
  discount_code?: string | null;
}

export interface WebAppRenewConfirmResponse {
  ok: boolean;
  message?: string | null;
  new_balance?: number | null;
  new_volume?: string | null;
  amount_paid?: number | null;
  config_name?: string | null;
  error?: string | null;
}
